packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js

```diff
@@ -0,0 +1,158 @@
+   *      Must be an array of at least one element.
+   * @param {Object?} denominatorPriceFeed We optionally divide the price spread between the spot and future baskets by this denominator price
+   *      in order to "denominate" the basket spread price in a specified unit. For example, we might want to express the basket spread in terms
+   *      of ETH-USD.
+   * @param {Number} lowerBoundSpread lower bound that the resultant value can take on
+   * @param {Number} upperBoundSpread upper bound that the resultant value can take on
+   */
+  constructor(web3, logger, spotPriceFeeds, futurePriceFeeds, denominatorPriceFeed, lowerBoundSpread, upperBoundSpread) {
+    super();
+
+    if (spotPriceFeeds.length === 0 || futurePriceFeeds.length === 0) {
+      throw new Error("BasisPriceFeed cannot be constructed with empty spot or future baskets.");
+    }
+
+    this.spotPriceFeeds = spotPriceFeeds;
+    this.futurePriceFeeds = futurePriceFeeds;
+    this.denominatorPriceFeed = denominatorPriceFeed;
+
+    this.lowerBoundSpread = lowerBoundSpread;
+    this.upperBoundSpread = upperBoundSpread;
+
+    // For convenience, concatenate all constituent price feeds.
+    this.allPriceFeeds = this.spotPriceFeeds.concat(this.futurePriceFeeds);
+    if (this.denominatorPriceFeed) {
+      this.allPriceFeeds = this.allPriceFeeds.concat(this.denominatorPriceFeed);
+    }
+
+    // Helper modules.
+    this.web3 = web3;
+    this.toBN = this.web3.utils.toBN;
+    this.toWei = this.web3.utils.toWei;
+    this.logger = logger;
+
+    // The precision that the user wants to return prices in must match all basket constituent price feeds and the denominator.
+    this.decimals = this.allPriceFeeds[0].getPriceFeedDecimals();
+
+    // Scale `number` by 10**decimals.
+    this.convertPriceFeedDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number
+      return this.toBN(parseFixed(number.toString(), this.decimals).toString());
+    };
+  }
+
+  // Given lists of future and spot prices, and a denominator price,
+  // return the spread price, which is:
+  // (avg(future) - avg(spot) + 1) / denominator
+  _getSpreadFromBasketPrices(futurePrices, spotPrices, denominatorPrice) {
+    // Compute future basket mean.
+    if (
+      futurePrices.length === 0 ||
+      futurePrices.some(element => element === undefined || element === null)
+    ) {
+      return null;
+    }
+    const futureMean = this._computeMean(futurePrices);
+
+    // Second, compute the average of the spot pricefeeds.
+    if (spotPrices.length === 0 || spotPrices.some(element => element === undefined || element === null)) {
+      return null;
+    }
+    const spotMean = this._computeMean(spotPrices);
+
+    if (!spotMean || !futureMean) return null;
+
+    let spreadValue = futureMean.sub(spotMean).div(spotMean).mul(this.convertPriceFeedDecimals("100"));
```

Comment:
just for reference, the umip formula: `min(max(100.0 * [1 + (F-S)/S], 75.0), 125.0)`
I dont see here where you are adding `1` to the spread value, maybe its not needed though im not sure. what do you think?

Reply:
great catch. the PR was updated after this implementation so was out of sync. making changes now.

Reply:
Done